### PR TITLE
Add back `Function::*_with_env(…)`

### PR DESCRIPTION
**Pull request by silwol**
Closes: #3064

# Description

# Review

- [x] Add a short description of the change to the CHANGELOG.md file

### CODE REVIEW

File: lib/api/src/sys/externals/function.rs [hunk 1]
```diff
@@ -983,8 +1098,30 @@ mod inner {
...
     }
 
+    /// Empty trait to specify the kind of `HostFunction`: With or
+    /// without an environment.
+    ///
+    /// This trait is never aimed to be used by a user. It is used by
+    /// the trait system to automatically generate the appropriate
+    /// host functions.
+    #[doc(hidden)]
+    pub trait HostFunctionKind {}
```
Comment (epilys) [inline lib/api/src/sys/externals/function.rs:1108]: Traits that are not meant to be implemented outside the crate should use the [sealed trait pattern](https://rust-lang.github.io/api-guidelines/future-proofing.html#c-sealed) to ensure this on compile-time

**Review by epilys — COMMENTED** [review]
Looks good to me, ready for merging after rebasing and optionally implementing the minor suggested change for the trait.

**Review by syrusakbary — APPROVED** [review]
Great work!

**Pull request merged**